crypt.go

```diff
@@ -237,6 +238,132 @@ func getPassword(sourceDevicePath, description string, reader io.Reader) (string
 	return askPassword(sourceDevicePath, "Please enter the "+description+" for disk "+sourceDevicePath+":")
 }
 
+// SnapModelChecker is used for verifying whether a Snap device model is
+// authorized to access the data on a volume unlocked by this package.
+type SnapModelChecker struct {
+	volumeName string
+	keyData    *KeyData
+	auxKey     AuxiliaryKey
+}
+
+// IsModelAuthorized indicates whether the supplied Snap device model is
+// authorized to access the data on the decrypted volume with the device
+// mapper name returned by VolumeName.
+func (c *SnapModelChecker) IsModelAuthorized(model SnapModel) (bool, error) {
+	return c.keyData.IsSnapModelAuthorized(c.auxKey, model)
+}
+
+// VolumeName is the device mapper name of the volume associated with this
+// SnapModelChecker.
+func (c *SnapModelChecker) VolumeName() string {
+	return c.volumeName
+}
+
+type activateWithKeyDataError struct {
+	k   *KeyData
+	err error
+}
+
+func (e *activateWithKeyDataError) Error() string {
+	return fmt.Sprintf("%s: %v", e.k.ID(), e.err)
+}
+
+func (e *activateWithKeyDataError) Unwrap() error {
+	return e.err
+}
+
+type keyDataAndError struct {
+	*KeyData
+	err error
+}
+
+type activateWithKeyDataState struct {
+	volumeName       string
+	sourceDevicePath string
+	activateOptions  []string
+	keyringPrefix    string
+
+	keys []*keyDataAndError
+
+	keyData *KeyData
+	auxKey  AuxiliaryKey
+}
+
+func (s *activateWithKeyDataState) errors() (out []*activateWithKeyDataError) {
+	for _, k := range s.keys {
+		if k.err == nil {
+			continue
+		}
+		out = append(out, &activateWithKeyDataError{k: k.KeyData, err: k.err})
+	}
+	return
```

Comment:
(nitpick) I'm not a fan of "naked" returns

Reply:
This is fixed now.